src/server/pkg/obj/microsoft_client.go

```diff
@@ -116,71 +125,80 @@ type microsoftWriter struct {
 	container  string
 	blob       string
 	blobClient storage.BlobStorageClient
+	buf        *bytes.Buffer
+	nBlocks    int
+	eg         errgroup.Group
+	err        error
 }
 
 func newMicrosoftWriter(client *microsoftClient, name string) (*microsoftWriter, error) {
-	// create container
-	_, err := client.blobClient.CreateContainerIfNotExists(client.container, storage.ContainerAccessTypePrivate)
-	if err != nil {
+	if _, err := client.blobClient.CreateContainerIfNotExists(client.container, storage.ContainerAccessTypePrivate); err != nil {
 		return nil, err
 	}
-
-	// create blob
-	err = client.blobClient.CreateBlockBlob(client.container, name)
-	if err != nil {
+	if err := client.blobClient.CreateBlockBlob(client.container, name); err != nil {
 		return nil, err
 	}
-
 	return &microsoftWriter{
 		container:  client.container,
 		blob:       name,
 		blobClient: client.blobClient,
+		buf:        bytes.NewBuffer(bufPool.GetBuffer()[:0]),
```

Comment:
What does `[:0]` do? I've never seen that in go before. It's not creating an empty slice, right?

Reply:
`s[:0]` creates a new slice which uses the same underlying storage as `s`, but has size `0`. This is useful because it allows us to use it as the underlying storage for a `bytes.Buffer` and have that buffer start writing from the beginning of the slice. If you simply did `bytes.NewBuffer(bufPool.GetBuffer())` then I'd get a `bytes.Buffer` that contained 100MB of `0`s and when I wrote to it, it would reallocate the slice and tack the write onto the end of those 100MB. Semantically this would work find if I just didn't initialize `buf`, since it would still be a buffer of size `0` so I could append to it until I hit 100MB and then write it like I do now, the difference is that it would need to reallocate the buffer several times during those writes, this way it never has to reallocate and we can reuse the buffers.